source/opt/constructs.h

```diff
@@ -0,0 +1,229 @@
+// Copyright (c) 2016 Google Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and/or associated documentation files (the
+// "Materials"), to deal in the Materials without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Materials, and to
+// permit persons to whom the Materials are furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Materials.
+//
+// MODIFICATIONS TO THIS FILE MAY MEAN IT NO LONGER ACCURATELY REFLECTS
+// KHRONOS STANDARDS. THE UNMODIFIED, NORMATIVE VERSIONS OF KHRONOS
+// SPECIFICATIONS AND HEADER INFORMATION ARE LOCATED AT
+//    https://www.khronos.org/registry/
+//
+// THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+// This file defines the language constructs for representing a SPIR-V
+// module in memory.
+
+#ifndef LIBSPIRV_OPT_CONSTRUCTS_H_
+#define LIBSPIRV_OPT_CONSTRUCTS_H_
+
+#include <functional>
+#include <vector>
+
+#include "spirv-tools/libspirv.h"
+#include "spirv/spirv.h"
+
+namespace spvtools {
+namespace opt {
+namespace ir {
+
+class Function;
+class Module;
+
+// Operand is a overloaded term. For better understanding, in this file, we
+// use payload to mean what "operand" is supposed to mean in the SPIR-V spec.
+// Operand itself is reserved to mean the real operands to an instruction,
+// excluding type ids and result ids.
+
+// A payload to an Inst. It can be a type id, a result id, or a real logical
+// operand.
+struct Payload {
```

Comment:
I think Payload is a worse term than Operand.  The things you're trying to represent really are all logical operands in the SPIR-V sense, even if it's result ID and type ID.
Note that the SPIR-V spec uses the term "operand" to mean a single word in an instruction after the opcode/wordcount word.  Within SPIRV-Tools I've used "logical operand" to mean "a named parameter to an instruction, but which might take multiple words to express" (including the cases of result ID and type ID.)

I'd much prefer you call this class Operand and be clear in the blurb for the class that it's a logical operand.